### PR TITLE
ISPN-9715 Update internal embedded listeners to be non blocking

### PR DESCRIPTION
**Pull request by wburns**
* Additional changes

https://issues.jboss.org/browse/ISPN-9715

### CODE REVIEW

File: query/src/main/java/org/infinispan/query/backend/SearchFactoryHandler.java [hunk 1]
```diff
@@ -120,11 +129,11 @@ boolean hasIndex(final Class<?> c) {
       return searchFactory.getIndexBindings().get(c) != null;
    }
 
-   private void handleClusterRegistryRegistration(final Class<?> clazz) {
+   private CompletionStage<Void> handleClusterRegistryRegistration(final Class<?> clazz) {
       if (hasIndex(clazz)) {
-         return;
+         return null;
```
Comment (karesti) [inline query/src/main/java/org/infinispan/query/backend/SearchFactoryHandler.java]: return a completion stage that contains null value instead of null ?

Comment (wburns) [reply query/src/main/java/org/infinispan/query/backend/SearchFactoryHandler.java]: The listener code allows for a null return which means it is already complete. It is a more performant way than returning a completed `CompletionStage`.

**Comment by wburns** [comment]
For some reason a change to pass the executor didn't get committed, will push in a bit here.

**Comment by wburns** [comment]
Pushed changes to fix compilation.

File: query/src/main/java/org/infinispan/query/backend/SearchFactoryHandler.java (hunk 1 above)
Comment (danberindei) [reply query/src/main/java/org/infinispan/query/backend/SearchFactoryHandler.java]: Is it really better than `completedNull()`? The return value is never used :)

**Comment by danberindei** [comment]
* Unused import
* CompletionStage return value is not used.
* How does this change affect HotRodMergeTest?

File: query/src/main/java/org/infinispan/query/backend/SearchFactoryHandler.java (hunk 1 above)
Comment (wburns) [reply query/src/main/java/org/infinispan/query/backend/SearchFactoryHandler.java]: The return value is indeed used in CacheNotifierImpl as null is a valid value for listeners as documented.

**Comment by wburns** [comment]
> * Unused import

Fixing

> * CompletionStage return value is not used.

Yes, it is. It is returned from the listener invocation.

> * How does this change affect HotRodMergeTest?

This fixes it because this test has locking that permanently exhausts 6 async threads. Our test suite only allows 6 async threads so no operation can fire on an async thread. By invoking the topology changed on the same thread when nothing is required it stays within the test limit of threads. When thread pools are reconfigured this will be a non issue.

**Comment by wburns** [comment]
> > * CompletionStage return value is not used.
> 
> Yes, it is. It is returned from the listener invocation.

Weird, sorry I must have missed pushing a commit again, which is probably why the import is messed up as well. Fixing.

**Comment by wburns** [comment]
Added in returns that was missing that also fixes the import :D

**Comment by wburns** [comment]
Updated a few more references to use persistence executor as a blocking executor, instead of async transport. The commits need to be merged together before integrating though.

File: query/src/main/java/org/infinispan/query/backend/SearchFactoryHandler.java (hunk 1 above)
Comment (danberindei) [reply query/src/main/java/org/infinispan/query/backend/SearchFactoryHandler.java]: It wasn't when I looked :)

   @Listener
   final class CacheListener {

      @CacheEntryCreated
      public void created(CacheEntryCreatedEvent<KeyValuePair<String, Class>, Boolean> e) {
         if (!e.isOriginLocal() && !e.isPre() && e.getValue()) {
            handleClusterRegistryRegistration(e.getKey().getValue());
         }
      }

      @CacheEntryModified
      public void modified(CacheEntryModifiedEvent<KeyValuePair<String, Class>, Boolean> e) {
         if (!e.isOriginLocal() && !e.isPre() && e.getValue()) {
            handleClusterRegistryRegistration(e.getKey().getValue());
         }
      }
   }

Comment (wburns) [reply query/src/main/java/org/infinispan/query/backend/SearchFactoryHandler.java]: I clarified on the PR directly.

**Comment by danberindei** [comment]
@wburns unused import again :(

**Comment by wburns** [comment]
Already fixed.

**Comment by wburns** [comment]
And turn that frown upside down :P

**Comment by danberindei** [comment]
Integrated, thanks Will!

**Pull request merged**